src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOverridingAndHiding.cs

```diff
@@ -1112,54 +1112,156 @@ public static void Main()
-    public override void Method(ref List<int> a) { Console.WriteLine(""Derived.Method(ref)""); } // No warning when ambiguous signatures are spread across multiple base types
+    public override void Method(ref List<int> a, out List<int> b)
+    {
+        b = null;
+        Console.WriteLine(""Derived.Method(ref List<int> a, out List<int> b)"");
+    }
+    public override void Method(BASEREF List<int> a)
+    {
+        a = null;
+        Console.WriteLine(""Derived.Method(BASEREF List<int> a)"");
+    }
 }
-class Test
+class Program
 {
-    public static void Main()
+    static void Main()
     {
-        Base<int, int> b = new Derived();
-        List<int> arg = new List<int>();
-        b.Method(out arg, ref arg);
-        b.Method(ref arg, out arg);
-        b.Method(ref arg);
+        List<int> t = null;
+        Console.WriteLine();
+        Derived d = new Derived();
+        d.Method(ref t, out t);
+        d.Method(out t, ref t);
+        d.Method(BASEREF t);
+        d.Method(BASE2REF t);
+        Console.WriteLine();
+        Base2<int, int> b2 = d;
+        b2.Method(ref t, out t);
+        b2.Method(out t, ref t);
+        b2.Method(BASEREF t);
+        b2.Method(BASE2REF t);
+        Console.WriteLine();
+        Base<int, int> b1 = d;
+        b1.Method(ref t, out t);
+        b1.Method(out t, ref t);
+        b1.Method(BASEREF t);
     }
 }
 ";
-            // Note: This test is exercising a case that is 'Runtime Ambiguous'. In the generated IL, it is ambiguous which
-            // method is being overridden. As far as I can tell, the output won't change from build to build / machine to machine
-            // although it may change from one version of the CLR to another (not sure). If it turns out that this makes
-            // the test flaky, we can delete this test.
-
-            var comp = CompileAndVerify(source, expectedOutput: @"
-Derived.Method(ref, out)
-Base.Method(ref, out)
-Base.Method(ref)");
+            for (int i = 0; i < 2; i++)
+            {
+                var substitutedSource0 = i switch
+                {
+                    0 => source.Replace("METHOD1", method1).Replace("METHOD2", method2),
+                    _ => source.Replace("METHOD2", method1).Replace("METHOD1", method2),
+                };
+                for (int j = 0; j < 2; j++)
+                {
+                    string subst(string s) => j switch
+                    {
+                        0 => s.Replace("BASEREF", "ref").Replace("BASE2REF", "out"),
+                        _ => s.Replace("BASEREF", "out").Replace("BASE2REF", "ref"),
+                    };
+
+                    var substitutedSource = subst(substitutedSource0);
+                    var compilation = CreateCompilation(substitutedSource, options: TestOptions.ReleaseExe);
+                    string expectedOutput;
+                    if (compilation.Assembly.RuntimeSupportsCovariantReturnsOfClasses)
```

Comment:
>if (compilation.Assembly.RuntimeSupportsCovariantReturnsOfClasses) [](start = 20, length = 66)

Once we have a fixed runtime available, I think it would be good to add an unconditional test against that runtime to verify expected behavior. Consider capturing this in the issue.

Reply:
Recorded in [https://github.com/dotnet/roslyn/issues/43188](https://github.com/dotnet/roslyn/issues/43188) for you.

---
In reply to: [446331200](https://github.com/dotnet/roslyn/pull/44025#discussion_r446331200) [](ancestors = 446331200)

Reply:
> Recorded in [https://github.com/dotnet/roslyn/issues/43188](https://github.com/dotnet/roslyn/issues/43188) for you.

I believe this test has nothing to do with covariant return scenario, and I am not asking to test that.

---
In reply to: [446411375](https://github.com/dotnet/roslyn/pull/44025#discussion_r446411375) [](ancestors = 446411375,446331200)

Reply:
The expected behavior for this scenario depends on a having the covariant-return supporting runtime where the necessary fix appears, which we do not yet have. For testing on existing runtimes, see line 1213.

---
In reply to: [446418074](https://github.com/dotnet/roslyn/pull/44025#discussion_r446418074) [](ancestors = 446418074,446411375,446331200)

Reply:
> The expected behavior for this scenario depends on a having the covariant-return supporting runtime where the necessary fix appears, which we do not yet have. For testing on existing runtimes, see line 1213.

That line tests conditionally. I am suggesting to add an unconditional test against core clr once the fix is going to be available to us.

---
In reply to: [446423506](https://github.com/dotnet/roslyn/pull/44025#discussion_r446423506) [](ancestors = 446423506,446418074,446411375,446331200)

Reply:
An unconditional test would fail on desktop. It would also fail on coreclr3. In any case, the request is recorded in [https://github.com/dotnet/roslyn/issues/43188](https://github.com/dotnet/roslyn/issues/43188).

---
In reply to: [446445016](https://github.com/dotnet/roslyn/pull/44025#discussion_r446445016) [](ancestors = 446445016,446423506,446418074,446411375,446331200)

Reply:
> An unconditional test would fail on desktop. It would also fail on coreclr3.

We are not testing against different flavors of Core CLR and I specifically asked for a test with runtime that includes the fix. We have capability to create targeted tests like that

---
In reply to: [446451821](https://github.com/dotnet/roslyn/pull/44025#discussion_r446451821) [](ancestors = 446451821,446445016,446423506,446418074,446411375,446331200)